src/network/connection.ts

```diff
@@ -105,63 +117,89 @@ export class Connection extends EventEmitter {
 
-    this.emit('connecting')
-    /* c8 ignore next 3 */
-    this.#socket = this.#options.tls
-      ? createTLSConnection(port, host, { ...this.#options.tls, ...connectionOptions })
-      : createConnection({ ...connectionOptions, port, host })
-    this.#socket.setNoDelay(true)
+        connectionsConnectsChannel.error.publish(diagnosticContext)
+        connectionsConnectsChannel.asyncStart.publish(diagnosticContext)
+        this.emit('timeout', error)
+        this.emit('error', error)
+        connectionsConnectsChannel.asyncEnd.publish(diagnosticContext)
+      }
 
-    this.#socket.once('connect', () => {
-      this.#socket.removeListener('timeout', connectionTimeoutHandler)
-      this.#socket.removeListener('error', connectionErrorHandler)
+      const connectionErrorHandler = (e: Error) => {
+        const error = new NetworkError(`Connection to ${host}:${port} failed.`, { cause: e })
+        diagnosticContext.error = error
 
-      this.#socket.on('error', this.#onError.bind(this))
-      this.#socket.on('data', this.#onData.bind(this))
-      this.#socket.on('drain', this.#onDrain.bind(this))
-      this.#socket.on('close', this.#onClose.bind(this))
+        this.#status = ConnectionStatuses.ERROR
 
-      this.#socket.setTimeout(0)
+        connectionsConnectsChannel.error.publish(diagnosticContext)
+        connectionsConnectsChannel.asyncStart.publish(diagnosticContext)
+        this.emit('error', error)
+        connectionsConnectsChannel.asyncEnd.publish(diagnosticContext)
+      }
 
-      this.#status = ConnectionStatuses.CONNECTED
-      this.emit('connect')
-    })
+      this.emit('connecting')
+      /* c8 ignore next 3 */
+      this.#socket = this.#options.tls
+        ? createTLSConnection(port, host, { ...this.#options.tls, ...connectionOptions })
+        : createConnection({ ...connectionOptions, port, host })
+      this.#socket.setNoDelay(true)
+
+      this.#socket.once('connect', () => {
+        this.#socket.removeListener('timeout', connectionTimeoutHandler)
+        this.#socket.removeListener('error', connectionErrorHandler)
+
+        this.#socket.on('error', this.#onError.bind(this))
+        this.#socket.on('data', this.#onData.bind(this))
+        this.#socket.on('drain', this.#onDrain.bind(this))
+        this.#socket.on('close', this.#onClose.bind(this))
+
+        this.#socket.setTimeout(0)
+
+        this.#status = ConnectionStatuses.CONNECTED
+
+        connectionsConnectsChannel.asyncStart.publish(diagnosticContext)
+        this.emit('connect')
+        connectionsConnectsChannel.asyncEnd.publish(diagnosticContext)
+      })
 
-    this.#socket.once('timeout', connectionTimeoutHandler)
-    this.#socket.once('error', connectionErrorHandler)
+      this.#socket.once('timeout', connectionTimeoutHandler)
+      this.#socket.once('error', connectionErrorHandler)
+      /* c8 ignore next 5 */
+    } catch (error) {
+      connectionsConnectsChannel.error.publish({ ...diagnosticContext, error })
```

Comment:
Should do context.error = error to keep reference equality.

Reply:
Updated!